### PR TITLE
Project restructuring, TCN models, and learning rate scheduling

### PR DESCRIPTION
**Pull request by npeard**
Bigly project restructuring. Implemented TCN models and learning rate scheduling, along with other minor improvements.

### CODE REVIEW

File: self_interferometry/main.py [hunk 1]
```diff
@@ -1,200 +1,195 @@
...
+        seed: Optional seed value. If None, a random seed will be generated.
+
+    Returns:
+        The seed value used.
+    """
+    if seed is None:
+        # Create a non-seeded RNG to generate a seed
+        temp_rng = np.random.default_rng()
+        # Generate a random seed between 0 and 2^32 - 1
+        seed = temp_rng.integers(0, 2**32 - 1)
```
Comment (ange1a-j14) [inline self_interferometry/main.py]: Why high exclusive on 2**32 - 1? More intuitive to use 2**32 so range is [0, 2^32 - 1]?

File: self_interferometry/redpitaya/coil_driver.py
Comment (ange1a-j14) [inline self_interferometry/redpitaya/coil_driver.py:1]: Line 108: could run into problems if t has length < 2 (even though unlikely). Could include a check? 

`get_velocity` and `get_displacement`: could implement a helper function `get_displacement_spectrum` to use in both, since there's a lot of repeat code. 

Line 307: In `integrate_velocity` header, could make default `sample_rate` a global constant since 125e6/256 is hardcoded in lots of different files. `sample_rate` is also a param in multiple functions but only has a default value in `integrate_velocity`. Is there any particular reason? 

Line 348: I believe the `for` loop over `batch_size` in `integrate_velocity` could be replaced by vectorization since each index is separate, which can make the code run more efficiently.

File: self_interferometry/redpitaya/manager.py
Comment (ange1a-j14) [inline self_interferometry/redpitaya/manager.py:1]: Line 654-5: Small change, but could use `-=` instead since the variable names are quite long, so switching to `-=` could improve readability. 

`save_data`: some confusion with `current_size` implementation. Given the update of `f.attrs['num_samples']` in line 754, why is `current_size` initialized as `f[first_key].shape[0]` in line 712 rather than using the `num_samples` attribute? 

`update_histograms`: could make # of bins a variable rather than hardcoding 100 each time. 
For the `if` statements of lines 1243, 1299, and 1352, could potentially consolidate into a `for` loop since the code is extremely similar (and long), just varying plot index, data/spectrum and color. 

Overall, could make pause of 0.01 a class variable since it's hardcoded throughout multiple functions.

File: self_interferometry/redpitaya/waveform.py
Comment (ange1a-j14) [inline self_interferometry/redpitaya/waveform.py:1]: Lines 23-4: Could import constants from RedPitayaManager instead of copy-pasting in case values change and want consistency. 

Lines 220-2: Given that `valid_freqs` is initialized in the constructor, if it is empty, wouldn't we want to throw an error (or at least raise a warning) rather than use the middle of the range since there would be incorrect behavior?  

Lines 261-6: some confusion about `test_mode` implementation. Just because we don’t want randomization if `test_mode` is true doesn’t necessarily mean if not `test_mode`, we always want randomization? You could potentially use an enum block to choose random phase only, random single tone, random all, or random none, since that seems to correspond more directly to the intention of the `if`/`elif` block. Or more simply, instead of using `not test_mode` to randomize in the `elif` block, you could use a parameter more directly detailing to randomize both spectrum and phase (s.t. setting it to false corresponds to `test_mode` behavior, but it isn’t defined by `test_mode`).

File: self_interferometry/signal_analysis/datasets.py
Comment (ange1a-j14) [inline self_interferometry/signal_analysis/datasets.py:1]: Line 78: Is there a reason you're using `len(f[self.voltage_key])` instead of `f.attrs[‘num_samples’]`? It doesn't seem like there's a strong need for `f.attrs[‘num_samples’]`. 

Line 98: Here is an example where you could have the default sample rate w/ dec=256 as a constant in RedPitayaManager and import it rather than hardcode it in multiple places. 

Line 144: I thought `__getitem__` is called multiple times when loading data, so I was confused why channel dropout is implemented in `__getitem__` since I'd assume it would only need to be done once.

File: self_interferometry/signal_analysis/generate_data.py
Comment (ange1a-j14) [inline self_interferometry/signal_analysis/generate_data.py:1]: Line 104: Another example where you could have the default sample rate w/ dec=256 as a constant in RedPitayaManager and import it rather than hardcode it in multiple places.

File: self_interferometry/signal_analysis/inspect_data.py
Comment (ange1a-j14) [inline self_interferometry/signal_analysis/inspect_data.py:1]: Line 36 and 131: Other examples where you could have the default sample rate w/ dec=256 as a constant in RedPitayaManager and import it rather than hardcode it in multiple places.

File: self_interferometry/signal_analysis/lightning_teacher.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            self.model_config.window_stride = 1
+
+        velocity_hat = self(signals)
+        displacement_hat = CoilDriver.integrate_velocity(velocity_hat)
+
+        return (
+            velocity_hat,
+            velocity_target,
+            displacement_hat,
+            displacement_target,
```
Comment (ange1a-j14) [inline self_interferometry/signal_analysis/lightning_teacher.py:285]: Did you want to shift `displacement_target` to start from 0 before returning, similar to in Standard's `predict_step`? Could also do with `displacement_hat` for consistency, but already done in CoilDriver’s `integrate_velocity`.

File: self_interferometry/signal_analysis/models_fcn.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        should be the same as the input size.
+
+        Args:
+            input_size: The length of the input sequence
+
+        Returns:
+            The length of the output sequence
+        """
+        # With proper padding in dilated convolutions, output size equals input size
+        return input_size
```
Comment (ange1a-j14) [inline self_interferometry/signal_analysis/models_fcn.py]: Even if `input_size` should match `output_size`, probably most direct to return `self.output_size` instead, just in case dimensions get mismatched, or there's some sort of error.

**Review by ange1a-j14 — COMMENTED** [review]
Overall, everything looks great with clear documentation. I added some file comments for files where I couldn't comment on a specific line, as well as 2 line comments.

File: self_interferometry/main.py (hunk 1 above)
Comment (npeard) [reply self_interferometry/main.py]: documentation indicates that the high value is inclusive: https://numpy.org/doc/2.2/reference/random/generated/numpy.random.Generator.integers.html
For 64bit floats, 2**32-1 is the largest positive int we can represent.

Comment (npeard) [reply self_interferometry/main.py]: Nevermind, read the docs incorrectly. Needs update to 2**32.

File: self_interferometry/redpitaya/coil_driver.py
Comment (npeard) [reply self_interferometry/redpitaya/coil_driver.py:1]: Line 108: Unnecessary, why would we ever use a single time point?

Implemented get_displacement_spectrum(), good idea.

Line 307: The default argument was included to allow Standard(L.LightningModule) to call integrate_velocity without specifying the sample rate, which is indeed obtuse. The proper thing to do would have Standard (in lightning_standard.py), get the sample rate stored in the data file, but this requires some restructuring of the hparams passing which I don't have time to get into right now. We can fix this once we start actually using some other sample rates (maybe with that RedPitayaDAQ code base). For now, I've moved the hardcoding of the sample rate to where it is used in the loss_function of Standard, which makes things a bit clearer. 

Line 348: Good catch. I'm training some models on the GPU right now, so I can't see how much this speeds things up, but I assume you are correct.

File: self_interferometry/redpitaya/manager.py
Comment (npeard) [reply self_interferometry/redpitaya/manager.py:1]: Line 654: OK, done.

save_data: Because if we are appending to an existing file, maybe f.attrs[] was modified and set incorrectly at some point - actually reading the size of the stored array is more explicit. 

update_histograms, num_bins: It's likely that the proper number of histogram for each type of plot is different. I've not had a problem with 100 yet, so I think this is a feature to add later when/if it becomes a problem.
if statements and pauses: Those if statements just check that the data exists before plotting. I'm actually not sure what good values for the pauses are, I think that needs some testing. Overall, this comment seems to me like a larger modification to RedPitayaManager than I can tackle right now. I agree there is a lot of repeated code in the update_histograms and update_plots that could be modularized and we should address. 

Can you raise a new Issue referencing that repeated code? Then one or both of us can make the fix (and do the testing) at a later point.

File: self_interferometry/redpitaya/waveform.py
Comment (npeard) [reply self_interferometry/redpitaya/waveform.py:1]: Lines 23-24: Yep.

Lines 220-222: OK.

Lines 261-6: The purpose of test_mode is to skip randomization of spectra when sample() has already been called before. I've renamed the parameter and updated the documentation. I'm not convinced yet that the current arguments are not sufficient or that a different set of arguments would be less confusing. You should check the updates and see if it makes more sense now.

File: self_interferometry/signal_analysis/datasets.py
Comment (npeard) [reply self_interferometry/signal_analysis/datasets.py:1]: Line 78: I don't see f.attrs['num_samples'] in this file. Not sure what you want.

Line 98: I think to make this optimal, we would have to refactor the use of RedPitayaManager.settings and perhaps move some or all of those settings to the newly created RedPitayaConfig data class. I'm not sure the best way to do that right now, it seems like we do want a mutable self.settings instance variable for some things (like start and end frequencies). Compromise for now is to have the hardware sample rate come from RedPitayaConfig but keep a hard-coded decimation (since this is a fallback case if h5_file.attrs fails, want it to be very clear where the self.sample_rate comes from, and avoid having to double check state of self.settings).

Line 144: No, you want to the dropped channel to change batch to batch. Otherwise, there would be some channels that the network would never see.

File: self_interferometry/signal_analysis/inspect_data.py
Comment (npeard) [reply self_interferometry/signal_analysis/inspect_data.py:1]: Implemented same compromise solution as above.

File: self_interferometry/signal_analysis/lightning_teacher.py (hunk 2 above)
Comment (npeard) [reply self_interferometry/signal_analysis/lightning_teacher.py:285]: This class is incomplete and needs further work before it is used, so I won't attempt edits now. We may not use it at all since my new opinion is that the self-supervised learning approach is going to be too complex and tedious to implement. If you want to work on the Teacher-Student approach, go for it.

File: self_interferometry/signal_analysis/models_fcn.py (hunk 3 above)
Comment (npeard) [reply self_interferometry/signal_analysis/models_fcn.py]: Probably. I'm just going to delete this class. I've not been using it and I don't think it's all that different from a TCN anyway (and we expect the TCN to do better anyway because of the dilation and causal windowing) nor do I have plans to test it further in the future. You can resurrect it and develop it further if you want.

File: self_interferometry/redpitaya/coil_driver.py [hunk 4]
```diff
@@ -314,7 +307,8 @@ def integrate_velocity(
         Args:
             velocity_waveform: Velocity waveform (microns/s)
                               Can be a NumPy array or PyTorch tensor
-                              For PyTorch tensors, supports batch dimensions [batch_size, channels, signal_length]
+                              For PyTorch tensors, supports batch dimensions
+                              [batch_size, channels, signal_length]
             sample_rate: Sample rate of the velocity waveform (Hz)
                         Default is Red Pitaya sample rate with decimation
```
Comment (ange1a-j14) [inline self_interferometry/redpitaya/coil_driver.py]: Can remove this line since removed default `sample_rate` value

File: self_interferometry/redpitaya/coil_driver.py [hunk 5]
```diff
@@ -314,7 +307,8 @@ def integrate_velocity(
...
             high_pass_freq: High-pass filter cutoff frequency (Hz) to remove DC drift
```
Comment (ange1a-j14) [inline self_interferometry/redpitaya/coil_driver.py]: `high_pass_freq` is also no longer a parameter

File: self_interferometry/redpitaya/coil_driver.py [hunk 6]
```diff
@@ -328,33 +322,15 @@ def integrate_velocity(
...
-                signal_length = original_shape[-1]
-
-                # Reshape for integration (flatten batch dimensions)
-                velocity_flat = velocity_waveform.reshape(-1, signal_length)
-                displacement_flat = torch.zeros_like(velocity_flat)
+            if len(velocity_waveform.shape) > 1:
+                batch_size, signal_length = velocity_waveform.shape
 
                 # Time step
                 dt = 1.0 / sample_rate
```
Comment (ange1a-j14) [inline self_interferometry/redpitaya/coil_driver.py]: Can pull out `dt = 1.0 / sample_rate` from lines 330, 337, and 348 because `sample_rate` is the same in all cases. Could just calculate `dt` once at the beginning of the function.

File: self_interferometry/redpitaya/redpitaya_config.py
Comment (ange1a-j14) [inline self_interferometry/redpitaya/redpitaya_config.py:1]: Could potentially have RedPitayaConfig dataclass in same file as CalibrationParameters dataclass? I'm not sure if they're related enough to be in the same file, but could be convenient while still readable.

File: self_interferometry/redpitaya/waveform.py [hunk 7]
```diff
@@ -251,7 +265,10 @@ def sample(
             self._random_single_tone_spectrum()
         elif randomize_phase_only:
             self._randomize_phase()
-        elif not test_mode:
+        elif not skip_randomization or self.spectrum is None:
```
Comment (ange1a-j14) [inline self_interferometry/redpitaya/waveform.py:268]: I really like this change and clear explanation!

File: self_interferometry/signal_analysis/lightning_standard.py [hunk 8]
```diff
@@ -145,55 +181,88 @@ def configure_optimizers(self) -> dict[str, Any]:
...
         """
         # Base velocity loss (MSE)
         velocity_loss = nn.MSELoss()(velocity_hat, velocity_target)
+        # Convert velocity loss units to um**2/ms**2 instead of um**2/s**2
+        # This allows SGD optimizer to run, apparently it can't handle large
+        # loss values.
+        velocity_loss /= 1e6
+
+        # Displacement loss (MSE)
+        sample_rate = 125e6 / 256
```
Comment (ange1a-j14) [inline self_interferometry/signal_analysis/lightning_standard.py]: Can use constants to calculate `sample_rate` here as well.

**Review by ange1a-j14 — COMMENTED** [review]
Some minor modifications possible, but overall looks great.

File: self_interferometry/redpitaya/coil_driver.py (hunk 4 above)
Comment (npeard) [reply self_interferometry/redpitaya/coil_driver.py]: DOne

File: self_interferometry/redpitaya/coil_driver.py (hunk 5 above)
Comment (npeard) [reply self_interferometry/redpitaya/coil_driver.py]: True

File: self_interferometry/redpitaya/coil_driver.py (hunk 6 above)
Comment (npeard) [reply self_interferometry/redpitaya/coil_driver.py]: True

File: self_interferometry/redpitaya/redpitaya_config.py
Comment (npeard) [reply self_interferometry/redpitaya/redpitaya_config.py:1]: No - usually try to go for one class/module per file. I don't think these are that closely related.

File: self_interferometry/redpitaya/waveform.py (hunk 7 above)
Comment (npeard) [reply self_interferometry/redpitaya/waveform.py:268]: Great

File: self_interferometry/signal_analysis/lightning_standard.py (hunk 8 above)
Comment (npeard) [reply self_interferometry/signal_analysis/lightning_standard.py]: True

File: self_interferometry/signal_analysis/generate_data.py
Comment (npeard) [reply self_interferometry/signal_analysis/generate_data.py:1]: Yes

**Review by ange1a-j14 — APPROVED** [review]
